zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -211,81 +211,92 @@ impl TransactionRecordsById {
         });
     }
 
+    fn find_sapling_spend(&self, nullifier: &sapling_crypto::Nullifier) -> Option<&SaplingNote> {
+        self.values()
+            .flat_map(|wallet_transaction_record| wallet_transaction_record.sapling_notes())
+            .find(|&note| {
+                if let Some(nf) = note.nullifier() {
+                    nf == *nullifier
+                } else {
+                    false
+                }
+            })
+    }
+    fn find_orchard_spend(&self, nullifier: &orchard::note::Nullifier) -> Option<&OrchardNote> {
+        self.values()
+            .flat_map(|wallet_transaction_record| wallet_transaction_record.orchard_notes())
+            .find(|&note| {
+                if let Some(nf) = note.nullifier() {
+                    nf == *nullifier
+                } else {
+                    false
+                }
+            })
+    }
+    // returns all sapling notes spent in a given transaction.
+    // will fail if a spend is not found when `fail_on_miss` is set to true.
     fn get_sapling_notes_spent_in_tx(
         &self,
-        query: &TransactionRecord,
+        query_record: &TransactionRecord,
+        fail_on_miss: bool,
     ) -> Result<Vec<&SaplingNote>, FeeError> {
-        query
-            .spent_sapling_nullifiers()
-            .iter()
-            .map(|nullifier| {
-                self.values()
-                    .flat_map(|wallet_transaction_record| wallet_transaction_record.sapling_notes())
-                    .find(|&note| {
-                        if let Some(nf) = note.nullifier() {
-                            nf == *nullifier
-                        } else {
-                            false
-                        }
-                    })
-                    .ok_or(FeeError::SaplingSpendNotFound(*nullifier))
-            })
-            .collect::<Result<Vec<&SaplingNote>, FeeError>>()
+        let mut sapling_spends: Vec<&SaplingNote> =
+            Vec::with_capacity(query_record.spent_sapling_nullifiers.len());
+
+        for nullifier in query_record.spent_sapling_nullifiers() {
+            if let Some(spend) = self.find_sapling_spend(nullifier) {
+                sapling_spends.push(spend);
+            } else if fail_on_miss {
+                return Err(FeeError::SaplingSpendNotFound(*nullifier));
```

Comment:
not sure about this error type... nothing to do with a fee, really
perhaps make the error type into Vec<&SaplingNote>... then you can eliminate the bool and vary the return code while still representing whether the search is complete.

my understanding is that this function will fail if called with the boolean on an incomplete set:: ie it will highlight an input we havent scanned the output for yet